esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+    default:
+        throw makeStringExceptionV(-1, "gateway %s: '%s' is not of the form \"[ vault-id ':' ] secret-name\"", name, identification);
+    }
+    if (!secret)
+        throw makeStringExceptionV(-1, "gateway %s: '%s' does not identify an 'esp' category secret", name, identification);
+    if (!isEmptyString(permission) && !secret->getPropBool(permission))
+        throw makeStringExceptionV(-1, "gateway %s: '%s' does not grant '%s' permission", name, identification, permission);
+    if (!secret->getProp("url", url.clear()) || url.isEmpty())
+        throw makeStringExceptionV(-1, "gateway %s: '%s' missing required 'url' property; credential-only secrets not supported", name, identification);
+    const char* username = secret->queryProp("username");
+    if (isEmptyString(username) && !secret->getPropBool("insecure"))
+        throw makeStringExceptionV(-1, "gateway %s: '%s' missing expected 'username' property; set 'insecure` property to 'true' if credentials are not required", name, identification);
+    const char* password = secret->queryProp("password");
+    if (isEmptyString(username) && password)
+        throw makeStringExceptionV(-1, "gateaay %s: '%s' invalid use of password without username", name, identification);
+    adjustURL(url, username, password);
+    gateway.setProp("@url", url);
+}
+
+void EsdlServiceImpl::resolveGatewayInlineURL(IPTree& gateway) const
+{
+    const char* name = gateway.queryProp("@name");
+    StringBuffer url(gateway.queryProp("@url"));
+    if (url.isEmpty())
+        throw makeStringExceptionV(-1, "gateway %s: missing  required '@url' property", name);
+    const char* username = gateway.queryProp("@username");
+    const char* password = gateway.queryProp("@password");
+    if (isEmptyString(username) && password)
+        throw makeStringExceptionV(-1, "gateway %s: invalid credentials - password without username", name);
+    StringBuffer decryptedPassword;
+    if (!isEmptyString(password))
+    {
+        decrypt(decryptedPassword, password);
+        password = decryptedPassword;
+    }
+    adjustURL(url, username, password);
+    gateway.setProp("@url", url);
+}
+
+void EsdlServiceImpl::adjustURL(StringBuffer& url, const char* username, const char* password) const
+{
+    StringBuffer scheme, usernameSink, passwordSink, host, port, path;
+    Utils::SplitURL(url, scheme, usernameSink, passwordSink, host, port, path);
+
+    if (scheme.length() > 0 && host.length() > 0)
+    {
+        StringBuffer userinfo;
+        if (!isEmptyString(username))
+        {
+            encodeUrlUseridPassword(userinfo, username);
+            if (password && *password) // TODO: remove '&& *password' to distinguish empty from omitted passwords
+            {
+                userinfo.append(':');
+                encodeUrlUseridPassword(userinfo, password);
+            }
+        }
+
+        url.clear();
+        url.append(scheme);
+        url.append("://");
+        if (userinfo.length()>0 )
```

Comment:
Extra space.  I also think the >0 on many of these ifs is a bit superfluous.  if (userinfo.length()) would suffice.